src/Adapter/Product/Validate/ProductValidator.php

```diff
@@ -133,6 +160,100 @@ private function validateShipping(Product $product): void
+        $this->validateProductLocalizedProperty($product, 'available_later', ProductConstraintException::INVALID_AVAILABLE_LATER);
+        $this->validateProductLocalizedProperty($product, 'available_now', ProductConstraintException::INVALID_AVAILABLE_NOW);
+    }
+
+    /**
+     * @param Product $product
+     *
+     * @throws ProductStockConstraintException
+     */
+    private function validateClassicStock(Product $product): void
+    {
+        // Depends on stock is only available in advanced mode
+        if ((bool) $product->depends_on_stock) {
+            throw new ProductStockConstraintException(
+                'You cannot perform this action when PS_ADVANCED_STOCK_MANAGEMENT is disabled',
+                ProductStockConstraintException::ADVANCED_STOCK_MANAGEMENT_CONFIGURATION_DISABLED
+            );
+        }
+
+        if ((bool) $product->advanced_stock_management) {
+            throw new ProductStockConstraintException(
+                'You cannot perform this action when PS_ADVANCED_STOCK_MANAGEMENT is disabled',
+                ProductStockConstraintException::ADVANCED_STOCK_MANAGEMENT_CONFIGURATION_DISABLED
+            );
+        }
+    }
+
+    /**
+     * @param Product $product
+     *
+     * @throws ProductPackConstraintException
+     * @throws ProductStockConstraintException
+     */
+    private function validateAdvancedStock(Product $product): void
+    {
+        if ((bool) $product->depends_on_stock && !(bool) $product->advanced_stock_management) {
+            throw new ProductStockConstraintException(
+                'You cannot perform this action when advanced_stock_management is disabled on the product',
+                ProductStockConstraintException::ADVANCED_STOCK_MANAGEMENT_PRODUCT_DISABLED
+            );
+        }
+
+        $this->checkPackStockType($product);
+    }
+
+    /**
+     * @param Product $product
+     *
+     * @throws ProductPackConstraintException
+     */
+    private function checkPackStockType(Product $product): void
+    {
+        // If the product doesn't depend on stock or is not a Pack no problem
+        if (!$product->depends_on_stock || !Pack::isPack($product->id)) {
+            return;
+        }
+
+        // Get pack stock type (or default configuration if needed)
+        $packStockType = $product->pack_stock_type;
+        if ($packStockType === Pack::STOCK_TYPE_DEFAULT) {
+            $packStockType = $this->defaultPackStockType;
+        }
+
+        // Either the pack has its own stock, or else ALL products from the pack must depend on the stock as well
+        if ($packStockType === Pack::STOCK_TYPE_PACK_ONLY || Pack::allUsesAdvancedStockManagement($product->id)) {
+            return;
+        }
+
+        throw new ProductPackConstraintException(
+            'You cannot link your pack to product stock because one of them has no advanced stock enabled',
```

Comment:
not a blocker, just a suggestion.
I think its better to format messages in different manner, like:
`Pack cannot be linked to product stock because one of them has no advanced stock enabled`.
The idea is to remove the "addressing" to user, because its not important, whether its "**you**" or "**me**" trying to link the pack :stuck_out_tongue: